src/ai/backend/test/templates/template.py

```diff
@@ -88,33 +129,46 @@ async def context(self) -> AsyncIterator[None]:
 
     @final
     async def run_test(self, exporter: TestExporter) -> None:
-        try:
-            # NOTE: self.context() should be an async context manager
-            async with self.context():  # type: ignore
-                await self._template.run_test(exporter)
-                await exporter.export_stage_done(self.name)
-        except BaseException as e:
-            await exporter.export_stage_exception(self.name, e)
-            raise
+        async with _apply_wrapper_templates(self._wrapper_templates, exporter):
+            try:
+                async with self.context():  # type: ignore
+                    await self._template.run_test(exporter)
+                    await exporter.export_stage_done(self.name)
+            except BaseException as e:
+                await exporter.export_stage_exception(self.name, e)
+                raise
 
 
 class SequenceTestTemplate(TestTemplate, ABC):
-    def __init__(self, templates: list[TestTemplate]) -> None:
+    def __init__(
+        self,
+        templates: list[TestTemplate],
+        wrapper_templates: list[WrapperTestTemplateProtocol] = [],
+    ) -> None:
         """
         Initialize the sequence template with a list of test templates.
+
         :param templates: The list of test templates to run in sequence.
+        :param wrapper_templates: Optional list of wrapper templates to apply before running the test code.
         """
         self._templates = templates
+        self._wrapper_templates = wrapper_templates
 
     @final
     async def run_test(self, exporter: TestExporter) -> None:
         """
         Run the test case by executing each template in sequence.
         :param exporter: The exporter to use for exporting test results.
         """
-        for template in self._templates:
+        async with _apply_wrapper_templates(self._wrapper_templates, exporter):
             try:
-                await template.run_test(exporter)
+                for template in self._templates:
+                    try:
+                        await template.run_test(exporter)
+                        await exporter.export_stage_done(self.name)
+                    except BaseException as e:
+                        await exporter.export_stage_exception(self.name, e)
+                        raise
```

Comment:
It seems the implementation is wrong. The completion or failure status in each template should only be performed within each template and should only be called once.